core/oracles/ethereum_oracle_verifier.go

```diff
@@ -0,0 +1,268 @@
+package oracles
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"code.vegaprotocol.io/vega/core/evtforward/ethcall"
+
+	"github.com/ethereum/go-ethereum"
+
+	"code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/core/validators"
+	"code.vegaprotocol.io/vega/logging"
+)
+
+var ErrDuplicatedEthereumCallEvent = errors.New("duplicated call event")
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/witness_mock.go -package mocks code.vegaprotocol.io/vega/core/oracles Witness
+type Witness interface {
+	StartCheck(validators.Resource, func(interface{}, bool), time.Time) error
+	RestoreResource(validators.Resource, func(interface{}, bool)) error
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/oracle_data_broadcaster_mock.go -package mocks code.vegaprotocol.io/vega/core/oracles OracleDataBroadcaster
+type OracleDataBroadcaster interface {
+	BroadcastData(context.Context, OracleData) error
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/eth_call_spec_source_mock.go -package mocks code.vegaprotocol.io/vega/core/oracles EthCallSpecSource
+type EthCallSpecSource interface {
+	GetSpec(id string) (EthCallSpec, error)
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/eth_contract_caller.go -package mocks code.vegaprotocol.io/vega/core/oracles ContractCaller
+type ContractCaller interface {
+	CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/eth_confirmations_mock.go -package mocks code.vegaprotocol.io/vega/core/oracles EthereumConfirmations
+type EthereumConfirmations interface {
+	Check(uint64) error
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/eth_call_spec_mock.go -package mocks code.vegaprotocol.io/vega/core/oracles EthCallSpec
+type EthCallSpec interface {
+	Normalise(callResult []byte) (map[string]string, error)
+	Call(ctx context.Context, caller ethereum.ContractCaller, blockNumber *big.Int) ([]byte, error)
+	PassesFilters(result []byte, blockHeight uint64, blockTime uint64) bool
+	RequiredConfirmations() uint64
+}
+
+type EthereumOracleVerifier struct {
+	log *logging.Logger
+
+	witness           Witness
+	timeService       TimeService
+	oracleEngine      OracleDataBroadcaster
+	ethCallSpecSource EthCallSpecSource
+	ethContractCaller ethereum.ContractCaller
+	ethConfirmations  EthereumConfirmations
+
+	pendingCallEvents    []*pendingCallEvents
+	finalizedCallResults []*types.EthContractCallEvent
+
+	mu     sync.Mutex
+	hashes map[string]struct{}
+
+	// snapshot data
+	snapshotState *ethereumOracleVerifierSnapshotState
+}
+
+type pendingCallEvents struct {
```

Comment:
Should this be `pendingCallEvent` singular? Seems weird to have the slice in the engine named the same thing as the struct `pendingCallEvents    []*pendingCallEvents`